hearth/lib/hearth/middleware/retry.rb

```diff
@@ -2,114 +2,63 @@
-        @client_rate_limiter = client_rate_limiter
-
         # instance state
-        @capacity_amount = nil
         @retries = 0
       end
 
-      # @param input
-      # @param context
-      # @return [Output]
-      # rubocop:disable Metrics/AbcSize
+      # rubocop:disable Metrics
       def call(input, context)
-        acquire_token
-        output = @app.call(input, context)
-        error_inspector = @error_inspector_class.new(
-          output.error, context.response.status
+        token = @retry_strategy.acquire_initial_retry_token(
+          context.metadata[:retry_token_scope]
         )
-        request_bookkeeping(output, error_inspector)
+        raise 'Unable to make request, too many failures' if token.nil?
 
-        return output unless retryable?(context, output, error_inspector)
+        output = nil
+        loop do
+          output = @app.call(input, context)
 
-        return output if @retries >= @max_attempts - 1
+          if (error = output.error)
+            error_info = @error_inspector_class.new(error, context.response)
+            token = @retry_strategy.refresh_retry_token(token, error_info)
 
-        @capacity_amount = @retry_quota.checkout_capacity(error_inspector)
-        return output unless @capacity_amount.positive?
+            Kernel.sleep(token.retry_delay) if token
+          else
+            @retry_strategy.record_success(token)
+          end
+          break unless token && output.error
 
-        delay = [Kernel.rand * (2**@retries), MAX_BACKOFF].min
-        Kernel.sleep(delay)
-        retry_request(input, context, output)
+          reset_request(context, output)
+        end
+        output
       end
-      # rubocop:enable Metrics/AbcSize
+      # rubocop:enable Metrics
 
       private
 
-      def acquire_token
-        return if @retry_mode == 'standard'
-
-        # either fail fast or block until a token becomes available
-        # must be configurable
-        # need a maximum rate at which we can send requests (max_send_rate)
-        # is unset until a throttle is seen
-        @client_rate_limiter.token_bucket_acquire(
-          1, wait_to_fill: @adaptive_retry_wait_to_fill
-        )
-      end
-
-      # max_send_rate is updated if on adaptive mode and based on response
-      # retry quota is updated if the request is successful (both modes)
-      def request_bookkeeping(output, error_inspector)
-        @retry_quota.release(@capacity_amount) unless output.error
-
-        return unless @retry_mode == 'adaptive'
-
-        @client_rate_limiter.update_sending_rate(
-          error_inspector.error_type == 'Throttling'
-        )
-      end
-
-      def retryable?(context, output, error_inspector)
-        return false unless output.error
-
-        error_inspector.retryable? &&
-          context.response.body.respond_to?(:truncate)
-      end
-
-      def retry_request(input, context, output)
+      def reset_request(context, output)
         @retries += 1
```

Comment:
It doesn't feel like this belongs in "reset_request" (maybe just move it outside of that method).

Reply:
It was previously in retry_request. I can move it out. Ideally I think we shouldn't need it either..